### PR TITLE
Fix for quaternion normalization while adding collision objects (Issue #1119) #rosindia #wmd

### PR DESCRIPTION
**Pull request by ggupta9777**
### Description

The quaternion message has been normalized before conversion to eigen for loading the object to the world. I checked it using the /get_planning_scene service as mentioned in the original issue. (1119). 

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Include a screenshot if changing a GUI
- [ ] Document API changes relevant to the user in the moveit/MIGRATION.md notes
- [ ] Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Decide if this should be cherry-picked to other current ROS branches
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_core/planning_scene/src/planning_scene.cpp [hunk 1]
```diff
@@ -1702,7 +1703,15 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
       if (s)
       {
         Eigen::Affine3d p;
-        tf::poseMsgToEigen(object.primitive_poses[i], p);
+        // To normalize any input quaternion
+        geometry_msgs::Pose primitive_pose = object.primitive_poses[i];
+        tf::Quaternion tf_quaternion;
+        tf::quaternionMsgToTF(object.primitive_poses[i].orientation, tf_quaternion);
```
Comment (BryceStevenWilley) [inline moveit_core/planning_scene/src/planning_scene.cpp]:
```suggestion
        tf::quaternionMsgToTF(primitive_pose, tf_quaternion);
```

File: moveit_core/planning_scene/src/planning_scene.cpp [hunk 2]
```diff
@@ -1702,7 +1703,15 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
...
+        tf_quaternion.normalize();
+        geometry_msgs::Quaternion msg_quaternion;
+        quaternionTFToMsg (tf_quaternion, msg_quaternion);
```
Comment (BryceStevenWilley) [inline moveit_core/planning_scene/src/planning_scene.cpp]:
```suggestion
        quaternionTFToMsg (tf_quaternion.normalized(), primitive_pose.orientation);
        // Delete extra lines above and below
```

File: moveit_core/planning_scene/src/planning_scene.cpp [hunk 3]
```diff
@@ -1702,7 +1703,15 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
...
       {
         Eigen::Affine3d p;
-        tf::poseMsgToEigen(object.primitive_poses[i], p);
+        // To normalize any input quaternion
+        geometry_msgs::Pose primitive_pose = object.primitive_poses[i];
+        tf::Quaternion tf_quaternion;
+        tf::quaternionMsgToTF(object.primitive_poses[i].orientation, tf_quaternion);
+        tf_quaternion.normalize();
+        geometry_msgs::Quaternion msg_quaternion;
+        quaternionTFToMsg(tf_quaternion, msg_quaternion);
```
Comment (BryceStevenWilley) [inline moveit_core/planning_scene/src/planning_scene.cpp]:
```suggestion
        quaternionTFToMsg(tf_quaternion.normalized(), primitive_pose.orientation);
        // Delete extra lines above and below
```

File: moveit_core/planning_scene/src/planning_scene.cpp [hunk 4]
```diff
@@ -1702,7 +1703,15 @@ bool PlanningScene::processCollisionObjectMsg(const moveit_msgs::CollisionObject
...
+        primitive_pose.orientation = msg_quaternion;
+        tf::poseMsgToEigen(primitive_pose, p);
```
Comment (v4hn) [inline moveit_core/planning_scene/src/planning_scene.cpp]: This is quite a bulk of code and I am pretty sure this is not the only place in this file where quaternions *should* be normalized.
Can you please look through the rest of the places that use quaternions here and wrap the validation into a function?

**Comment by ggupta9777** [comment]
Thanks for the feedback, I'll make it more compact first and then look at where else this can be implemented.

**Comment by henningkayser** [comment]
@ggupta9777 do you think you can find time to finish this PR?

**Comment by ggupta9777** [comment]
@henningkayser Yes, I'll finish this one by the weekend.

**Comment by henningkayser** [comment]
> @henningkayser Yes, I'll finish this one by the weekend.

That's great, thanks!

**Comment by ggupta9777** [comment]
> That's great, thanks!

I have run into some problems with clang and compile warning, it will take me some more time to get to it again. It's available for take up if someone else you or someone else interested in finishing it earlier.

**Comment by felixvd** [comment]
@ggupta9777 What are the issues you ran into? Do you need help getting over a bump?

**Comment by j-petit** [comment]
I think I did this in #1420.

**Comment by rhaschke** [comment]
Closing, as this was handled in #1420.

**Pull request closed without merging**